File: qna/src/main/java/com/example/rag/qna/adapter/output/opinion/OpinionPersistenceAdapter.java
package com.example.rag.qna.adapter.output.opinion;

import com.example.rag.qna.domain.Opinion;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.List;

@Component
@AllArgsConstructor
public class OpinionPersistenceAdapter implements OpinionPersistencePort {
    private final OpinionRepository opinionRepository;

    @Override
    public void saveOpinion(Opinion o) {
        opinionRepository.save(
                new OpinionEntity(o.getQuestionId(), o.getTitle(), o.getContent(), o.getUserId(), LocalDateTime.now(), LocalDateTime.now()));
    }

    @Override
    public void updateOpinion(Opinion opinion) {
        OpinionEntity entity = getOpinionEntity(opinion.getOpinionId());
        entity.setTitle(opinion.getTitle());
        entity.setContent(opinion.getContent());
        entity.setUpdatedAt(LocalDateTime.now());
    }

    @Override
    public void deleteOpinion(String opinionId) {
        opinionRepository.deleteById(getOpinionIdForMySQL(opinionId));
    }

    @Override
    public Opinion findOpinionById(String opinionId) {
        return toOpinion(getOpinionEntity(opinionId));
    }

    @Override
    public List<Opinion> findOpinionsByQuestionId(String questionId, int offset, int limit) {
        return opinionRepository.findByQuestionId(questionId, offset, limit).stream()
                .map(OpinionPersistenceAdapter::toOpinion)
                .toList();
    }

    private OpinionEntity getOpinionEntity(String opinionId) {
        return opinionRepository.findById(getOpinionIdForMySQL(opinionId)).orElseThrow();
    }

    private static Long getOpinionIdForMySQL(String opinionId) {
        return opinionId == null ? null : Long.parseLong(opinionId);
    }

    private static Opinion toOpinion(OpinionEntity entity) {
        return new Opinion(
                getOpinionIdForTransparency(entity.getId()),
                entity.getQuestionId(),
                entity.getTitle(),
                entity.getContent(),
                entity.getUserId(),
                entity.getCreatedAt(),
                entity.getUpdatedAt()
        );
    }

    private static String getOpinionIdForTransparency(Long opinionId) {
        return String.valueOf(opinionId);
    }
}

File: qna/src/main/kotlin/com/example/rag/qna/adapter/output/opinion/OpinionPersistenceAdapter.kt
package com.example.rag.qna.adapter.output.opinion

import core.rag.Opinion
import org.springframework.stereotype.Component
import java.time.LocalDateTime


@Component
class OpinionPersistenceAdapter(private val opinionRepository: OpinionRepository) : OpinionPersistencePort {
    private companion object OpinionMapper {
        fun idForTransparency(opinionId: Long?) = opinionId.toString()
        fun idForMySQL(opinionId: String?) = opinionId?.toLong() ?: -1

        fun Opinion.createEntity() = OpinionEntity(questionId, title, content, userId, LocalDateTime.now())
        fun OpinionEntity.toDomainModel() = Opinion(idForTransparency(id), questionId, title, content, userId, createdAt, updatedAt)
        fun OpinionEntity.updateWithDomain(domain: Opinion) {
            this.title = domain.title
            this.content = domain.content
            this.updatedAt = LocalDateTime.now()
        }
    }

    override fun saveOpinion(domain: Opinion): OpinionEntity = opinionRepository.save(domain.createEntity())
    override fun updateOpinion(domain: Opinion) = getOpinionEntity(domain.questionId).updateWithDomain(domain)
    override fun deleteOpinion(opinionId: String): Unit = opinionRepository.deleteById(idForMySQL(opinionId))
    override fun findOpinionById(opinionId: String): Opinion = getOpinionEntity(opinionId).toDomainModel()
    override fun findOpinionsByQuestionId(questionId: String, offset: Int, limit: Int): List<Opinion> = getOpinions(questionId, offset, limit)

    private fun getOpinionEntity(opinionId: String?): OpinionEntity = opinionRepository.findById(idForMySQL(opinionId)).orElseThrow()
    private fun getOpinions(questionId: String, offset: Int, limit: Int) = opinionRepository.findByQuestionId(questionId, offset, limit)
                                                                                            .map { entity -> entity.toDomainModel() }
}
